httpservice/http_service.go

```diff
@@ -0,0 +1,148 @@
+package httpservice
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/remind101/pkg/retry"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+// HTTPService provides a way to make http calls with an aggressive transport
+type HTTPService interface {
+	Call(method string, path string, jsonData interface{}) ([]byte, error)
+	CallWithRetries(method string, path string, jsonData interface{}) ([]byte, error)
+}
+
+// HTTPClient contains the retrier, baseURL and the actual httpClient.
+type HTTPClient struct {
+	client      *http.Client
+	baseURL     *url.URL
+	retrier     *retry.Retrier
+	serviceName string
+}
+
+// NewHTTPClient returns a new HTTPClient for the given service name and the base url
+// The baseURL will be used to prefix the paths provided in Call and CallWithRetries
+func NewHTTPClient(serviceName string, baseURL *url.URL) *HTTPClient {
+	aggressiveTransport := &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+		Dial: (&net.Dialer{
+			Timeout:   1 * time.Second, // !!!
+			KeepAlive: 90 * time.Second,
+		}).Dial,
+		TLSHandshakeTimeout: 3 * time.Second,
+	}
+
+	return &HTTPClient{
+		baseURL: baseURL,
+		client: &http.Client{
+			Transport: aggressiveTransport,
+		},
+		retrier: retry.NewErrorTypeRetrier(serviceName,
+			retry.DefaultBackOffOpts,
+			(*net.OpError)(nil),
+			(*RetryableHTTPError)(nil)),
+	}
+}
+
+// CallWithRetries makes calls to the network with the default retry logic
+func (client *HTTPClient) CallWithRetries(method string, path string, jsonData interface{}) ([]byte, error) {
+	res, err := client.retrier.Retry(func() (interface{}, error) {
+		return client.Call(method, path, jsonData)
+	})
+	if err != nil {
+		return []byte{}, err
+	}
+	return res.([]byte), nil
+}
+
+// Call formats the network request and returns the raw data as the response
+func (client *HTTPClient) Call(method string, path string, jsonData interface{}) ([]byte, error) {
+	req, err := client.buildJSONRequest(method, client.parseURL(path), jsonData)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := client.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode >= 500 {
+		return nil, &RetryableHTTPError{Path: path, StatusCode: resp.StatusCode}
+	} else if resp.StatusCode >= 300 {
+		return nil, &HTTPError{Path: path, StatusCode: resp.StatusCode}
+	}
+
+	return ioutil.ReadAll(resp.Body)
+}
+
+func (client *HTTPClient) parseURL(path string) string {
+	baseURL := client.urlWithoutCreds(*client.baseURL)
```

Comment:
the whole `urlWithoutCreds` business prevents basic auth credentials from showing up in stack traces. i think we want to keep this.